pkg/controller/wildflyserver/wildflyserver_controller.go

```diff
@@ -253,41 +248,66 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
-		statefuleSet.Spec.Template.Spec.Containers[0].Env = env
-		update = true
-	}
-	// Ensure the envFrom variables are up to date
-	envFrom := w.Spec.EnvFrom
-	if !reflect.DeepEqual(statefuleSet.Spec.Template.Spec.Containers[0].EnvFrom, envFrom) {
-		log.Info("Updating envFrom", "StatefulSet.Namespace", statefuleSet.Namespace, "StatefulSet.Name", statefuleSet.Name)
-		statefuleSet.Spec.Template.Spec.Containers[0].EnvFrom = envFrom
-		update = true
+// checkStatefulSet checks if the statefulset is up to date with the current WildFlyServerSpec.
+// it returns true if a reconcile result must be returned.
+// the 2nd boolean specifies whether the result must be required.
+// A non-nil error if an error happend while updating/deleting the statefulset.
+func (r *ReconcileWildFlyServer) checkStatefulSet(wildflyServer *wildflyv1alpha1.WildFlyServer, foundStatefulSet *appsv1.StatefulSet) (bool, bool, error) {
+	if generationStr, found := foundStatefulSet.Annotations["wildfly.org/wildfly-server-generation"]; found {
+		if generation, err := strconv.ParseInt(generationStr, 10, 64); err == nil {
+			// WildFlyServer spec has possibly changed, delete the statefulset
+			// so that a new one is created from the updated spec
+			if generation != wildflyServer.Generation {
+				statefulSet := r.statefulSetForWildFly(wildflyServer)
+				delete := false
+				// changes to VolumeClaimTemplates can not be updated and requires a delete/create of the statefulset
+				if len(statefulSet.Spec.VolumeClaimTemplates) > 0 {
+					if len(foundStatefulSet.Spec.VolumeClaimTemplates) == 0 {
+						// existing stateful set does not have a VCT
+						delete = true
+					} else {
+						foundVCT := foundStatefulSet.Spec.VolumeClaimTemplates[0]
+						vct := statefulSet.Spec.VolumeClaimTemplates[0]
+
+						if foundVCT.Name != vct.Name ||
+							!reflect.DeepEqual(foundVCT.Spec.AccessModes, vct.Spec.AccessModes) ||
+							!reflect.DeepEqual(foundVCT.Spec.Resources, vct.Spec.Resources) {
+							delete = true
+						}
+					}
+				} else {
+					if len(foundStatefulSet.Spec.VolumeClaimTemplates) != 0 {
+						// existing stateful set has a VCT while updated statefulset does not
+						delete = true
+					}
+				}
+
+				if delete {
+					// VolumeClaimTemplates has changed, the statefulset can not be updated and must be deleted
+					if err = r.client.Delete(context.TODO(), foundStatefulSet); err != nil {
+						log.Error(err, "Failed to Delete StatefulSet.", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
+						return true, false, err
+					}
+					log.Info("Deleting StatefulSet that is not up to date with the WildFlyServer StorageSpec", "StatefulSet.Namespace", foundStatefulSet.Namespace, "StatefulSet.Name", foundStatefulSet.Name)
+					return true, true, nil
+				}
+
+				// all other changes are in the spec Template or Replicas and the statefulset can be updated
+				foundStatefulSet.Spec.Template = statefulSet.Spec.Template
```

Comment:
@jmesnil : just checking as I don't understand here if the changes will be reflected in pods

As the template definition is base for the pod creation, does the changing of template cause changing of the currently active pods? Does not need the spec template change removing the statefulset and recreating it?